### PR TITLE
Add tests for mutability

### PR DESCRIPTION
**Pull request by dkcumming**
Adding tests to examine the boundaries of mutability and immutability, not all of these use `unsafe` but only the ones I listed as unsafe do we annotate unsafe directly:
- mut_const.rs: A UB test that mutates the data under an immutable value through pointer arithmetic
- interior-mut-fail.rs: Safely has interior mutability through `RefCell` (stuck on `Rvalue::AddressOf`)
- interior-mut2-fail.rs: Safely has interior mutability through `Cell` (stuck on `Rvalue::AddressOf`)
- interior-mut3-fail.rs: Unsafe but sound interior mutability through `UnsafeCell` (stuck on `Rvalue::AddressOf`)

We discussed the possibility of adding tests that show demonstrate immutable field of mutable structs, but this is not possible in rust. There is a notion of private fields if structs come from other modules, but this is not the same as mutability.

### CODE REVIEW

File: kmir/src/tests/integration/test_integration.py [hunk 1]
```diff
@@ -436,9 +436,7 @@ def test_prove(spec: Path, tmp_path: Path, kmir: KMIR) -> None:
 
 PROVING_DIR = (Path(__file__).parent / 'data' / 'prove-rs').resolve(strict=True)
 PROVING_FILES = list(PROVING_DIR.glob('*.rs'))
-PROVE_RS_SHOW_SPECS = [
-    'local-raw-fail',
-]
+PROVE_RS_SHOW_SPECS = ['local-raw-fail', 'interior-mut-fail', 'interior-mut2-fail', 'interior-mut3-fail']
```
Comment (ehildenb) [inline kmir/src/tests/integration/test_integration.py:439]: We should do these one per line instead. Will make it easier to add/delete them.

**Comment by ehildenb** [comment]
Can we rebase this on master to get current `master` behavior? I want to open a PR with the refactoring from https://github.com/runtimeverification/mir-semantics/issues/573), and make sure we all agree on the changes that happen then.

File: kmir/src/tests/integration/test_integration.py (hunk 1 above)
Comment (dkcumming) [reply kmir/src/tests/integration/test_integration.py:439]: Can do

**Review by gtrepta — APPROVED** [review]
